Feature/braen 04/product service dev

SmesiteJl opened the pull request:
1. Categories CRUD ✅
2. Products CRUD ✅
3. Logs ✅
4. Tests ✅
5. JDBC ✅
6. Validation ✅
7. Swagger ✅

SmesiteJl committed:
initial commit

initial commit w/ product minimal crud, swagger and validation ready

SmesiteJl committed:
base working version

Stable version with a given package architecture. Jdbc implementation for product, swagger w/o custom OpenAPI, base migrations and validation are ready

SmesiteJl committed:
all w/o tests and one endpoint

1. Categories CRUD ✅/❌
2. Products CRUD ✅
3. Logs ✅
4. Tests ❌
5. JDBC ✅
6. Validation ✅
7. Swagger ✅

SmesiteJl committed:
runnable now

E-Line21 commented on services/product-service/.gitignore:
Мне кажется этот .gitignore не нужен. Это всё есть в корневом .gitignore

E-Line21 commented on services/product-service/product-api/build.gradle:
```diff
@@ -0,0 +1,11 @@
+}
+
+dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-web'
```
Лучше вместо стартера подключи только зависимости с нужными тебе интерфейсами и аннотациями, без автоконфигурации. Для того, кто будет подключать данный модуль, чтобы слать клиентские запросы, может стать неожиданностью, что у него еще появится автоконфигурация

E-Line21 commented on services/product-service/product-api/src/main/java/com/technokratos/eateasy/product/api/CategoryApi.java:
```diff
@@ -0,0 +1,79 @@
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "Categories retrieved successfully",
+                    content = @Content(schema = @Schema(implementation = CategoryResponse.class))),
+            @ApiResponse(responseCode = "404", description = "No categories found")
```
Лучше возвращать просто пустой список, а не 404, если нет категорий

E-Line21 commented on services/product-service/product-api/src/main/java/com/technokratos/eateasy/product/api/CategoryApi.java:
```diff
@@ -0,0 +1,79 @@
+    @GetMapping("/{id}/products")
+    ResponseEntity<List<ProductResponse>> getProductsByCategory(
+            @Parameter(description = "Category ID", example = "1", required = true)
+            @PathVariable Long id,
```
Исправь `example="1"`, у тебя id категории `UUID`. Соответственно, поменяй и тип параметра с `Long` на `UUID`

E-Line21 commented on services/product-service/product-api/src/main/java/com/technokratos/eateasy/product/api/CategoryApi.java:
```diff
@@ -0,0 +1,79 @@
+            @RequestParam(required = false) String page,
+
+            @Parameter(description = "Number of products per page", example = "10")
+            @RequestParam(required = false) String page_size,
```
Добавь валидацию, чтобы не было отрицательных или слишком больших значений

E-Line21 commented on services/product-service/product-api/src/main/java/com/technokratos/eateasy/product/dto/product/ProductRequest.java:
```diff
@@ -0,0 +1,43 @@
+        @NotEmpty(message = "Product category cannot be empty")
+        @Size(max = 255, message = "Product category is longer than 255 characters")
+        @Schema(description = "Product category", example = "Fruits")
+        List<UUID> categories,
```
Перепроверь аннотации для валидации. Как минимум в `@Shema` `example` точно не "Fruits". В целом не уверен, что оно будет отрабатывать так, как ожидается

E-Line21 commented on services/product-service/product-db/build.gradle:
```diff
@@ -0,0 +1,32 @@
+            url 'jdbc:postgresql://localhost:5432/product'
+            username 'eat_easy'
+            password 'easy'
+            driver 'org.postgresql.Driver'
```
Нужно сделать так, чтобы параметры бд (url, username, password и т.д.) предоставалялись извне

E-Line21 commented on services/product-service/product-db/db/changelog/db.changelog-master.yaml:
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  - include:
+      file: v1.0/v1.0-create_table_products.yaml
```
Добавь `relativeToChangelogFile: true`

E-Line21 commented on services/product-service/product-impl/build.gradle:
```diff
@@ -0,0 +1,30 @@
+tasks.withType(JavaCompile) {
+    options.annotationProcessorPath = configurations.annotationProcessor
+    options.compilerArgs = ['-Amapstruct.defaultComponentModel=spring']
+}
```
Добавь 
```
tasks.named('test') {
    useJUnitPlatform()
}
```

E-Line21 commented on services/product-service/product-api/src/main/java/com/technokratos/eateasy/product/api/CategoryApi.java:
```diff
@@ -0,0 +1,79 @@
+            @ApiResponse(responseCode = "404", description = "No categories found")
+    })
+    @GetMapping
+    ResponseEntity<List<CategoryResponse>> getAll();
```
Использование `ResponseEntity` здесь излишне. Лучше возвращать просто DTO + `@ResponseBody` + `@ResponseStatus`

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/controller/CategoryController.java:
```diff
@@ -0,0 +1,54 @@
+@Slf4j
+public class CategoryController implements CategoryApi {
+
+    private final CategoryService categoryService;
```
Достаточно назвать переменную просто `service`. Понятно, что `CategoryController` это сервис категорий

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/entity/Category.java:
Зачем в слое бизнес логики аннотации от Swagger?

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/repository/CategoryRepository.java:
Данная реализация имеет ряд проблем:
- Аннотация `@Transactional` не будет работать, т.к. каждый раз открывается новое соединение
- Оборачивать `SQLException` в `RuntimeException` смело, но это выстрел себе в колено.

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+
+    public Optional<List<Category>>getCategoriesByProductId(UUID productId) {
```
Зачем возвращать `Optional`? Можно просто вернуть пустой список

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/service/ProductService.java:
```diff
@@ -0,0 +1,104 @@
+                    .toEntity(product)));
+            log.info("Updated product with id: {}", id);
+            if(affectedRows == 0) {
+                log.warn("Product with id: {} not found", id);
```
Это не уровень `warn`, а скорее `info` или `debug`. `Warn` - это ошибка, которая не обрывает обработку текущего запроса, но в будущем может привести к проблемам. А в данном случае данная ошибка является штатной

SmesiteJl committed:
dry-pre release commit w/ tests

E-Line21 commented on services/product-service/product-impl/build.gradle:
```diff
@@ -0,0 +1,58 @@
+//tasks.register('copyMigrations', Copy) {
+//    from project(':services:product-service:product-db').file('db/changelog')
+//    into "${layout.buildDirectory.dir("resources/test").get()}/db/changelog"
+//}
```
Забыл комментарий

E-Line21 commented on services/product-service/product-db/db/changelog/db.changelog-master.yaml:
```diff
@@ -0,0 +1,4 @@
+databaseChangeLog:
+  - include:
+      file: db/changelog/v1.0/v1.0-create_table_products.yaml
+      elativeToChangelogFile: false
```
Опечатка

E-Line21 commented on services/product-service/product-api/src/main/java/com/technokratos/eateasy/product/api/CategoryApi.java:
```diff
@@ -0,0 +1,90 @@
+            @PathVariable("id") UUID id,
+
+            @RequestParam(name = "order_by", required = false, defaultValue = "popularity")
+            //@Pattern(regexp = "^(new|price|popularity)$", message = "order_by must be one of: new, price, popularity")
```
Комментарий забыл

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/controller/CategoryController.java:
```diff
@@ -0,0 +1,49 @@
+            Integer page,
+            Integer page_size,
+            BigDecimal max_price,
+            BigDecimal min_price
```
Подобное именование переменных в мире Java уголовно наказуемо😂

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/entity/Product.java:
А зачем ты указываешь аннотации для валидации над сущностями? Я чего-то не понимаю?

SmesiteJl replied:
Это я перестарался и забыл убрать, смысла в этом нет, согласен

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/repository/ProductRepository.java:
```diff
@@ -0,0 +1,139 @@
+        } catch (DataIntegrityViolationException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new RuntimeException("Database error", e);
```
Я не уверен, что стоит бросать именно RuntimeException. @robl0xx что думаешь?

rbrmnv replied:
Думаю стоит оставить RuntimeException, все таки это инфраструктурная ошибка, здесь код просто упадет

SmesiteJl replied:
Я посоветовался с умным другом, мне понравился его ответ, прикреплю его сюда, как обоснование 👇

> ### ✅ Почему именно `RuntimeException`:
> 
> 1. **Это стандартный подход в Spring, JDBC и JPA**  
>     Spring Framework **всё оборачивает в runtime-исключения** — `JdbcTemplate`, `JPA`, `RestTemplate`, даже `TransactionManager`. Почему? Потому что `checked`-исключения **нагружают слои приложения**.
>     
> 2. **Business vs Technical**
>     
>     - `checked` — если вызывающий код может и должен что-то сделать с ошибкой (например, показать пользователю, что данные невалидны).
>         
>     - `unchecked` — если ошибка говорит: "Здесь что-то пошло фундаментально не так, мы ничего не можем сделать, кроме как отвалиться красиво".
>         
> 3. **Пробрасывание `checked`-исключения нарушает SRP**  
>     Например, сервис-слой не должен знать, что репозиторий использует JDBC, если мы выбрасываем `SQLException` или кастомное `CheckedDbException` — это **нарушение инкапсуляции**.
>     
> 4. **Глобальный `@ControllerAdvice` будет обрабатывать это централизованно**  
>     Ты не теряешь контроль — наоборот: ты даёшь системе возможность централизованно логировать и обрабатывать критические сбои.

> При этом DataIntegrityViolationException, как бизнес-связанная, остаётся отдельно — её мы и должны обрабатывать осознанно.

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/repository/ProductRepository.java:
```diff
@@ -0,0 +1,139 @@
+    public int update(UUID productId, Map<String, Object> updates) {
+        if (updates.isEmpty()) return 0;
+
+        Set<String> allowedColumns = Set.of("title", "description", "photo_url", "price", "category", "quantity");
```
Можешь вынести в отдельное поле, чтобы каждый раз не создавать новый объект в куче?

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/repository/ProductRepository.java:
```diff
@@ -0,0 +1,139 @@
+        for (Map.Entry<String, Object> entry : updates.entrySet()) {
+            String column = entry.getKey();
+            if (!allowedColumns.contains(column)) {
+                throw new IllegalArgumentException("Invalid column name: " + column);
```
Я понимаю, что это исключение, вероятно, никогда не будет выброшено, но все же, конкатенация строк - это плохо

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/controller/ProductController.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public void update(UUID id, @RequestBody ProductRequest product) {
```
<img width="1297" alt="image" src="https://github.com/user-attachments/assets/a66bf926-d19f-436d-ace7-afcd01a17700" />

Ручка отрабатывает некорректно. Записать с таким id точно есть

E-Line21 commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/controller/CategoryController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public List<ProductResponse> getProductsByCategory(
```
<img width="1297" alt="image" src="https://github.com/user-attachments/assets/aa6db890-b06f-449a-877c-24e0989d0c8f" />

Тут явно неверное сообщение об ошибке

SmesiteJl committed:
fixed issues from pull request comments

fixed issues from pull request comments  w/o:
- "Тут явно неверное сообщение об ошибке"

robl0xx commented on services/build.gradle:
Зачем писать версии если у тебя есть dependency-management?

rbrmnv commented on services/product-service/product-api/src/main/java/com/technokratos/eateasy/product/api/ProductApi.java:
```diff
@@ -0,0 +1,122 @@
+            @ApiResponse(responseCode = "404", description = "Product not found")
+    })
+    @PatchMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @ResponseBody
+    void update(
```
Почему на UPDATE возвращаешь HttpStatus.NO_CONTENT?

SmesiteJl replied:
HTTP-метод `PUT` или `PATCH`, когда успешно выполняется без возврата тела, по стандарту RFC 7231 может вернуть:
* `200 OK` — если возвращается тело (например, обновлённый объект)
* `204 No Content `— если ничего не возвращается

У меня метод update - void, поэтому 204 тут как раз подходит

rbrmnv commented on services/product-service/product-db/gradle.properties:
```diff
@@ -0,0 +1,7 @@
+db.driver=org.postgresql.Driver
+db.changelogFile=db.changelog-master.yaml
+
+#here should be empty
```
Коммент

rbrmnv commented on services/product-service/product-db/gradle.properties:
Вообще этот файл не должен отслеживаться гитом

rbrmnv commented on services/product-service/product-impl/build.gradle:
```diff
@@ -0,0 +1,55 @@
+    implementation 'org.mapstruct:mapstruct:1.5.5.Final'
+    annotationProcessor 'org.mapstruct:mapstruct-processor:1.5.5.Final'
+
+    implementation 'org.projectlombok:lombok:1.18.36'
+    annotationProcessor 'org.projectlombok:lombok:1.18.36'
```
Возможно заработает и без annotationProcessor 'org.projectlombok:lombok:1.18.36', попробуй

rbrmnv commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/mapper/CategoryMapper.java:
А для update метода где нужно обновить только часть параметров он сработает?

SmesiteJl replied:
Да, проверил ещё раз, сработает ✅

rbrmnv commented on services/product-service/product-impl/src/main/resources/query/category/save.sql:
```diff
@@ -0,0 +1,3 @@
+INSERT INTO category (title)
+VALUES (?)
+RETURNING *;
```
Поясни за RETURNING *;

SmesiteJl replied:
Чтобы после сохранения новой сущности сразу получить обратно эту сущность и вернуть в ответе клиенту.
`RETURNING` — это SQL-оператор PostgreSQL, который позволяет немедленно получить данные строк, затронутых командой INSERT, UPDATE или DELETE, без необходимости выполнять отдельный SELECT-запрос после.

rbrmnv commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/service/ProductService.java:
Надо бы интерфейс сделать, а это типо его реализация

rbrmnv commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/service/ProductService.java:
```diff
@@ -0,0 +1,121 @@
+                            .toEntity(product)));
+        } catch (DataIntegrityViolationException e) {
+            log.error("Duplicate product {}", product);
+            throw new ProductAlreadyExistsException("Product already exists");
```
Теряешь исходную причину ошибки (e не передаётся)

rbrmnv commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/service/ProductService.java:
```diff
@@ -0,0 +1,121 @@
+            }
+        } catch (DataIntegrityViolationException e) {
+            log.error("Invalid product data with id {}", id);
+            throw new DataIntegrityViolationException("Data integrity violation");
```
Теряешь исходную причину ошибки (e не передаётся)

rbrmnv commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/service/ProductService.java:
```diff
@@ -0,0 +1,121 @@
+    @Transactional
+    public void update(UUID id, ProductUpdateRequest product) {
+        try {
+            int affectedRows = productRepository.update(id, prepareUpdatesMap(productMapper
+                    .toEntity(product)));
```
Не нравится мне маппер, как будто update ломаться будет

SmesiteJl replied:
Нет, проверил ещё раз. Не ломается. А что вызывает подозрения, @robl0xx ?

rbrmnv commented on services/product-service/product-impl/src/main/java/com/technokratos/eateasy/product/repository/ProductRepository.java:
```diff
@@ -0,0 +1,139 @@
+        String sql = queryProvider.getSqlQueryForProduct("update_quantity");
+        try {
+            return jdbcTemplate.update(sql, productId, quantity);
+        } catch (DataIntegrityViolationException e) {
+            throw e;
```
Я думаю этот блок можно убрать, он же ничего по сути не делает

SmesiteJl replied:
для маня важно отдельно выкидывать DataIntegrityViolationException, потому что он перехватывается обработчиком и выводит кастомное сообщение об ошибке. Дальше у меня перехватывается любая другая ошибка, они мне уже не важны с точки зрения бизнес-логики.

SmesiteJl committed:
typo corrected

SmesiteJl committed:
correct error message and codestyle

SmesiteJl committed:
Update build.gradle

SmesiteJl committed:
Interfaces injection added

SmesiteJl committed:
Merge branch 'develop' into feature/BRAEN-04/product-service-dev

SmesiteJl committed:
validation fix
